src/Features/CSharp/Portable/UseIsNullCheck/CSharpUseIsNullCheckForCastAndEqualityOperatorCodeFixProvider.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseIsNullCheck
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    internal class CSharpUseIsNullCheckForCastAndEqualityOperatorCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.UseIsNullCheckForCastAndEqualityOperatorDiagnosticId);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+
+            context.RegisterCodeFix(
+                new MyCodeAction(CSharpFeaturesResources.Use_is_null_check,
+                c => this.FixAsync(context.Document, diagnostic, c)),
+                context.Diagnostics);
+            return SpecializedTasks.EmptyTask;
+        }
+
+        protected override Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var binary = (BinaryExpressionSyntax)diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken: cancellationToken);
+                var rewritten = Rewrite(binary);
+
+                editor.ReplaceNode(binary, rewritten);
```

Comment:
Indeed I'm seeing this crash with the ReferenceEquals:
```c#
        if (object.ReferenceEquals(object.ReferenceEquals("", null), null))
        {
        }
```
"fix all" crashes here. Consider fixing that as well and adding a test for this to both features.

Reply:
> callback overload not needed? 

Not really.

> will this work with nested fixes?

Unlikely.  but also not very relevant as that's not a pattern i would expect to see in the wild, nor would i think it important that it be fixed.  Fix all (with the callback variant) is importnat when there is a realistic chance of these things actually being nested.

Reply:
So are you saying it's ok to crash on code like this 😕 the fix is really simple... I don't understand why we should ship crashing code fixes?

Reply:
> Please fix and add tests for this to both features.

I've addressed the crash in the existing feature, but i haven't make the fix actually do the fix-all here.  It seems oo unlikely to happen in practice, so having to try to handle htis case seems unnecessary.  I have just ensured that it at least won't crash those.

Reply:
> Why did you use this approach by ordering the diagnostics? It seems a little fragile. 

It's not fragile.  It's an approach we take in lots of other places when there is the chance that you may need to process an inner node within an outer node.

> here should be an overload of ReplaceNode that takes a callback which would make this scenario work

I'm fine if you want to spend cycles on this.  I don't think it's worthwhile to spend any more time on this given the 0% chance it will be hit by customer :)

Reply:
Honestly, using the callback version should be default and if you don't want to spend any effort, you shouldn't spend effort thinking whether you can afford to *not* use it. I've seen at least 3 different code fix crashes because the wrong overload was used here.

Reply:
the callback form doesn't prevent crashes.  ReplaceNode, today, fundamentally can just crash on you if you end up replacing a node that no longer exists due to a previous operation.  

i.e. say you had:

```Outer(Inner(a))```

And you do a ReplaceNode on 'Outer' (with some other node that does not include 'Inner'), then ReplaceNode on 'Inner' can crash on you regardless of which form you use (since you'll get null passed to your callback, which certainly nearly no callbacks support).

The important question is: is this a reasonable scenario where you could expect nesting or not?  In my opinion, the ```== null``` scenario is not.  So spending time on the callback form just isn't relevant.  

--

Note: i have long thought ReplaceNode should not have this behavior.  I raised that point several times over the lifetime of it, but was unable to get the change made.  The crash was felt to be helpful because it helped call out to the person that there was something wrong with their ordering, and having the editor just ignore the subsequent replacements was not appropriate.

Reply:
> And you do a ReplaceNode on 'Outer' (with some other node that does not include 'Inner'), 

Yes, but the replacement still includes the inner node. I believe using the callback form should make everything work.